Add docs for GCP Dataproc deployment (#4393)
kedro-org/kedro: docs/source/deployment/gcp_dataproc.md

# GCP Dataproc

> `Dataproc serverless` lets you run Spark workloads without requiring you to provision and manage your own Dataproc cluster. An advantage over `Dataproc compute engine` is that `Dataproc serverless` supports custom containers allowing you package your dependencies at build time. Refer [here](https://cloud.google.com/dataproc-serverless/docs/overview#s8s-compared) for the official comparison between Dataproc serverless and compute engine.

Reviewer: ```suggestion
`Dataproc Serverless` lets you run Spark workloads without requiring you to provision and manage your own Dataproc cluster. An advantage over `Dataproc compute engine` is that `Dataproc Serverless` supports custom containers allowing you package your dependencies at build time. Refer [here](https://cloud.google.com/dataproc-serverless/docs/overview#s8s-compared) for the official comparison between Dataproc Serverless and compute engine.
```

The guide details kedro pipeline deployment steps for `Dataproc serverless`.

## Overview

> The below diagram details the dataproc serverless dev and prod deployment workflows. 

Reviewer: ```suggestion
The below sections and diagrams detail the dataproc serverless dev and prod deployment workflows. 
```

### DEV deployment (and experimentation)

> The following are the steps:

Reviewer: ```suggestion
The following steps are needed to do a DEV deployment on Dataproc Serverless:
```

3. **Submit spark batches**: The developer triggers an experimental run with dataproc batches submission python script with parameters
   - `--env=dev`
   - `--run-id=...`
   - `--kedro-run-args=...`
3.1. **Prepare and push artifacts to GCS**: The script packages the active code changes in the branch into an archive, prepares any necessary artifacts and pushes to GCS in `{branch}/{run_id}` namespace.
3.2. **Pull docker image to dataproc**: Dataproc pulls the custom docker image tagged with the branch name from artifact registry
3.3. **Pull artifact archives**: Dataproc pulls the archives containing the artifacts and unzips them on master and all workers in the working directory
3.4. **Run pyspark job**: The main file then triggers the pyspark job as the standard kedro cli command: `kedro run --pipeline=... --params=...`
3.5. *(Optional)* **Run output**: It is recommended to segregate the experimental run outputs into `{branch}/{run_id}` namespace.
 

![GCP DEV deployment workflow](../meta/images/gcp_dataproc_serverless_dev_workflow.png)

### PROD deployment

> The following are the steps:

Reviewer: ```suggestion
The following steps are needed to do a PROD deployment on Dataproc Serverless:
```

1. **Cut a release from develop**: A release branch is cut from the `develop` branch as `release/v0.2.0`
2. **Prepare release**: Minor fixes, final readiness and release notes are added to prepare the release.

> 3. **Merge into main**: After all checks passes and necessary approvals, the release branch is merged into main, and the commit is tagged with the version

Reviewer: ```suggestion
3. **Merge into main**: After all checks pass and necessary approvals are received, the release branch is merged into main, and the commit is tagged with the version
```

4. **Deploy docker image**: The docker image is built with release tag version `v0.2.0` and pushed to prod artifact registry. Note the following:
  - `libraries` - Packaged
  - `artifacts` - Packaged
  - `code` - Packaged
  - `entrypoint.py` - Packaged
5.1. **Prepare and push artifacts to GCS**: The script prepares any necessary artifacts and pushes to GCS in `{branch}/{run_id}` namespace.
5.2. **Pull docker image to dataproc**: Dataproc pulls the custom docker image tagged with the branch name from artifact registry
5.3. **Pull artifact archives**: Dataproc pulls the archives containing the artifacts and unzips them on master and all workers in the working directory
5.4. **Run pyspark job**: The main file then triggers the pyspark job as the standard kedro cli command: `kedro run --pipeline=... --params=...`
5.5. *(Optional)* **Run output**: It is recommended to segregate the experimental run outputs into `{branch}/{run_id}` namespace.

![GCP PROD deployment workflow](../meta/images/gcp_dataproc_serverless_prod_workflow.png)

**NOTE**: The above describes a simple reference deployment pattern. Please adapt it for your usecase.

## Prerequisite Setup

> #### Create service accounts

Reviewer 2: ```suggestion
### Create service accounts
```

NOTE:

> 1. The service account creation method below assigns all permissions needed for this walkthrough in one service account. 

> > 2. Different teired environments may have their own GCP Projects.

Reviewer: ```suggestion
> 2. Different tiered environments may have their own GCP Projects.
```

    --iam-account=sa-dataproc-runner@${PROJECT_ID}.iam.gserviceaccount.com
```

> NOTE: The service account JSON keys can be stored as secrets, and can be made accessible to CI in your custom workflows.

### Create GCS Buckets

The following GCS Buckets are created for development (`env=dev`) and production (`env=prod`) workflows.

#### Dataproc temporary staging bucket

```bash
gcloud storage buckets create gs://dataproc-staging-${REGION}-${UID}-${ENV} --location ${REGION}
```

#### Dataproc run submit artifacts bucket

```bash
gcloud storage buckets create gs://dataproc-run-submit-${REGION}-${UID}-${ENV} --location ${REGION}
```

```bash
dataproc-run-submit-{REGION}-{UID}-dev/
└── {branch}/
    └── {run_id}/
```

```bash
dataproc-run-submit-{REGION}-{UID}-prod/
└── release/
    └── {version}/ # for e.g. v.2.0 
```

#### Dataproc run output bucket

```bash
gcloud storage buckets create gs://dataproc-run-output-${REGION}-${UID}-${ENV} --location {REGION}
```

```bash
dataproc-run-output-{REGION}-{UID}-dev/
└── {branch}/
    └── {run_id}/
```

```bash
dataproc-run-output-{REGION}-{UID}-prod/
└── release/
    └── {version}/ # for e.g. v.2.0 
        └── {run_id}
```

> #### Authorize with service account

Reviewer: Nit: we use British spelling in the Kedro docs 🤓 

```suggestion
#### Authorise with service account
```

WORKDIR /home/kedro

# Install and setup uv
RUN pip install --upgrade pip uv build
ENV UV_INDEX_STRATEGY=unsafe-any-match

# Install dependencies
COPY src/requirements.txt /tmp/requirements.txt
RUN uv pip install --system \
    -r /tmp/requirements.txt \
    && rm /tmp/requirements.txt
    

RUN chown -R kedro:${KEDRO_GID} /home/kedro

USER kedro

# development image
FROM base as dev

# production image
FROM base as prod

# copy source code
COPY . /code

RUN chmod -R a+w /home/kedro

```

### Configure Artifact registry

This creates your repository in the artifact registry.

```bash
gcloud artifacts repositories create ${ARTIFACT_REPOSITORY} \
    --repository-format=docker \
    --location=${REGION} \
    --description="Kedro project docker repository"
```

#### Gcloud configure docker auth

This configures gcloud authentication with the artifact registry.

```bash
gcloud auth configure-docker ${ARTIFACT_REGISTRY}
```

#### Push Image to Artifact Registry

`deployment/dataproc/serverless/build_push_docker.sh`

- This script builds and pushes the docker image for user dev workflows by tagging each custom build with the branch name (or a custom tag).

> - The developer can experiment with any customizations to the docker image in their feature branches.

Reviewer: ```suggestion
- The developer can experiment with any customisations to the docker image in their feature branches.
```
